frame/pablo/src/uniswap_tests.rs

```diff
@@ -428,3 +431,166 @@ proptest! {
+					base_price_cumulative: 1,
+					quote_price_cumulative: 1,
+					timestamp: 0,
+					base_twap: Rate::one(),
+					quote_twap: Rate::one()
+				})
+			);
+
+			update_max_price();
+
+			run_to_block(TWAP_INTERVAL + 1);
+			// execute a swap to invoke update_twap()
+			let usdt_value = 1_u128 * unit;
+			assert_ok!(Tokens::mint_into(USDT, &BOB, usdt_value));
+			assert_ok!(Pablo::swap(
+				Origin::signed(BOB),
+				pool_id,
+				CurrencyPair::new(BTC, USDT),
+				usdt_value,
+				0,
+				false
+			));
+
+			update_max_price();
+			// TWAP get updated
+			let elapsed = (TWAP_INTERVAL + 1) * MILLISECS_PER_BLOCK;
+			let twap = Pablo::twap(pool_id).expect("twap not found");
+			assert_eq!(twap.timestamp, elapsed);
+			// TWAP must be smaller than max price observed during the interval
+			assert!(twap.base_twap < max_base_price);
+			assert!(twap.quote_twap < max_quote_price);
+		});
+	}
+
+	#[test]
+	fn twap_asset_prices_change_within_twap_interval() {
+		new_test_ext().execute_with(|| {
+			let unit = 1_000_000_000_000_u128;
+			let initial_btc = 100_u128 * unit;
+			let initial_usdt = 100_u128 * unit;
+			let pool_id =
+				create_pool(BTC, USDT, initial_btc, initial_usdt, Permill::zero(), Permill::zero());
+
+			let mut min_base_price = Rate::from_float(99999999.0);
+			let mut min_quote_price = Rate::from_float(99999999.0);
+			let mut max_base_price = Rate::from_float(0.0);
+			let mut max_quote_price = Rate::from_float(0.0);
+			let mut update_min_max_price = || {
+				let base_price =
+					Pablo::do_get_exchange_rate(pool_id, crate::PriceRatio::NotSwapped);
+				assert_ok!(base_price);
+				let base_price = base_price.unwrap();
+				let quote_price = Pablo::do_get_exchange_rate(pool_id, crate::PriceRatio::Swapped);
+				assert_ok!(quote_price);
+				let quote_price = quote_price.unwrap();
+				min_base_price = sp_std::cmp::min(base_price, min_base_price);
+				min_quote_price = sp_std::cmp::min(quote_price, min_quote_price);
+				max_base_price = sp_std::cmp::max(base_price, max_base_price);
+				max_quote_price = sp_std::cmp::max(quote_price, max_quote_price);
+			};
+			System::set_block_number(0);
+			assert_eq!(Pablo::twap(pool_id), None);
+			assert_ok!(Pablo::enable_twap(Origin::root(), pool_id));
+			run_to_block(1);
+
+			assert_eq!(
+				Pablo::twap(pool_id),
+				Some(TimeWeightedAveragePrice {
+					base_price_cumulative: 1,
+					quote_price_cumulative: 1,
+					timestamp: 0,
+					base_twap: Rate::one(),
+					quote_twap: Rate::one()
+				})
+			);
+			let run_to_block_and_swap = |block_number: BlockNumber| {
+				run_to_block(block_number);
+				// execute a swap to invoke update_twap() on 5 moment
+				let usdt_value = 1_u128 * unit;
+				assert_ok!(Tokens::mint_into(USDT, &BOB, usdt_value));
+				assert_ok!(Pablo::swap(
+					Origin::signed(BOB),
+					pool_id,
+					CurrencyPair::new(BTC, USDT),
+					usdt_value,
+					0,
+					false
+				));
+			};
+			run_to_block_and_swap(5);
+			update_min_max_price();
```

Comment:
Shouldn't we test that the cumulative prices are updating correctly based on the swaps after these steps?